Kashoo/synctos
# Issue 313: Regular expression pattern constraint for attachment filenames

# Description

When included in a document definition, the `filenameRegexPattern` constraint verifies that all attachments for that document type conform to the specified regular expression pattern.

# Testing

1. Generate a sync function from `attachment-constraints-doc-definitions.js`:

```bash
./make-sync-function -j test/resources/attachment-constraints-doc-definitions.js build/test-sync-function.txt
```

2. Copy and paste the sync function into a Sync Gateway configuration file.
3. Create a test document:

```
PUT /test/attachment-test-1 HTTP/1.1
Host: localhost:4985
Content-Type: application/json

{
	"type": "staticAttachmentFilenameRegexPatternDoc"
}
```

4. Add an attachment with a valid filename to the document (it should succeed with 201 Created):

```
PUT /test/attachment-test-1/foo.xls?rev=1-35244b689bcb92f9767971da257112e8 HTTP/1.1
Host: localhost:4985
Content-Type: text/plain

This isn't actually an Excel document, despite the file extension, but it doesn't matter because the document type doesn't validate content type of attachments.
```

5. Attempt to add an attachment with an invalid filename (it should be rejected with 403 Forbidden):

```
PUT /test/attachment-test-1/bar.invalid?rev=2-6ddfb9f095264f2658ba82f5ed48e6bc HTTP/1.1
Host: localhost:4985
Content-Type: text/plain

The file name doesn't match the regex: /^(foo|bar)\.(xls|xlsx)$/
```

# Related Issue

* GitHub issue link: #313
